### ISSUE TITLE
Displaying errors from K8S to Jupyter Notebook

### ISSUE THREAD

**Issue by gogasca** [jupyter/enterprise_gateway#974]
## Description
I'm testing JEG with Kubernetes in Google Cloud.
When I start a new Kernel and this Kernel includes a GPU. There are cases in which no GPU is available and I would like to indicate that to the end user.
When i look into my Jupyter Notebook logs I get:

```
599 POST /api/sessions?1624229228607 (::1): Error attempting to connect to Gateway server url 'http://x.x.x.x:8888'.  Ensure gateway url is valid and the Gateway instance is running.
[W 2021-06-20 15:48:39.415 ServerApp] Error attempting to connect to Gateway server url 'http://x.x.x.x:8888'.  Ensure gateway url is valid and the Gateway instance is running.
[E 2021-06-20 15:48:39.418 ServerApp] {
      "Host": "localhost:8081",
      "Connection": "keep-alive",
      "Content-Length": "129",
      "Sec-Ch-Ua": "\" Not;A Brand\";v=\"99\", \"Google Chrome\";v=\"91\", \"Chromium\";v=\"91\"",
      "X-Xsrftoken": "2|1c14e8ec|37b8b4fc2b4c470117543c4574f8d222|1624176701",
      "Sec-Ch-Ua-Mobile": "?0",
      "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36",
      "Content-Type": "text/plain;charset=UTF-8",
      "Accept": "*/*",
      "Origin": "http://localhost:8081",
      "Sec-Fetch-Site": "same-origin",
      "Sec-Fetch-Mode": "cors",
      "Sec-Fetch-Dest": "empty",
      "Referer": "http://localhost:8081/lab/workspaces/auto-i/tree/gpu.ipynb",
      "Accept-Encoding": "gzip, deflate, br",
      "Accept-Language": "en-US,en;q=0.9,es;q=0.8",
      "Cookie": "username-localhost-8889=\"2|1:0|10:1622003661|23:username-localhost-8889|44:NzZiYjg4MTlkMGMxNDVhMGFkMmM4Yjg1Mjg0YTZhN2Q=|e4e4d853a29d75594501df37e485f3c5f5cd91134b337be066d284949b9ebab1\"; session=.eJw1kMtqxDAMRf_F64HYVvzKr5QhyLI0GZomg-0sSum_17R0KXTv0UFfapXKbVNLrxff1PosalEOiojWPuaYtUuMBSkL5pgipETGJoNAPs7WOtFZQhGwHlJ2oCVAAJAibBJlpuDZDlYIJMaXIui1D6RddkWjKzw7nYCQTAB0JM5GADVEXlw_8OCj_6tRq7L2852PYZgCyjByDueIHpHRDFuBMIdoaRy0MYjNdpAKPtbWsV9tlefeuY467vvY7CfhzmMcyJt64YPX7dn6WT_V8qa23l_LNP2GtrP1Jeqop0Gr1zHtIzep-01djevf04z6_gFpaWaz.YL-XBg.Jqb3WIKoonnvaZRSbzKDEJIUB-Y; _xsrf=2|1c14e8ec|37b8b4fc2b4c470117543c4574f8d222|1624176701"
    }

```

My question is around the handling of errors from GKE and its propagation to users in Jupyter Notebook.
1) Is there an "switch" config that I need to tweak to pass down some of these errors vs a 500 error?
2) During my initial testing of JEG, I did see sometimes JEG able to detect errors (Like no Docker container available, permissions downloading the container, etc.)
 
I'm willing to contribute to a thin layer which can propagate error back to user to indicate more information. 
Just would like to understand some of the existing status and architecture around these cases.
Right now the best I got was Error attempting to connect to Gateway server url 'http://x.y.x.x:8888' which operationally is not very helpful.

## Screenshots / Logs
If applicable, add screenshots and/or logs to help explain your problem.
To generate better logs, please run the gateway with `--debug` command line parameter.

## Environment

- Enterprise Gateway Version 2.5.0
- Platform:Kubernetes
- Others [e.g. Jupyter Server 5.7, JupyterHub 1.0, etc]

```
jupyter --version
jupyter core     : 4.7.1
jupyter-notebook : 6.3.0
qtconsole        : 5.0.3
ipython          : 7.22.0
ipykernel        : 5.3.4
jupyter client   : 6.1.12
jupyter lab      : 3.0.11
nbconvert        : 6.0.7
ipywidgets       : 7.6.3
nbformat         : 5.1.3
traitlets        : 5.0.5
```

**Comment by kevin-bates** [comment]
Hi @gogasca - the Jupyter framework is not very conducive for conveying errors that occur during a kernel's launch.  For example, I just "adjusted" a kernelspec to reference a non-existent file for its process and it produced the following output - w/o EG - just pure Notebook/Jupyter server:
```
      File "/opt/anaconda3/envs/elyra-dev/lib/python3.6/site-packages/jupyter_client/manager.py", line 661, in start_kernel
        self.kernel = await self._launch_kernel(kernel_cmd, **kw)
      File "/opt/anaconda3/envs/elyra-dev/lib/python3.6/site-packages/jupyter_client/manager.py", line 642, in _launch_kernel
        res = launch_kernel(kernel_cmd, **kw)
      File "/opt/anaconda3/envs/elyra-dev/lib/python3.6/site-packages/jupyter_client/launcher.py", line 132, in launch_kernel
        proc = Popen(cmd, **kwargs)
      File "/opt/anaconda3/envs/elyra-dev/lib/python3.6/subprocess.py", line 729, in __init__
        restore_signals, start_new_session)
      File "/opt/anaconda3/envs/elyra-dev/lib/python3.6/subprocess.py", line 1364, in _execute_child
        raise child_exception_type(errno_num, err_msg, err_filename)
    FileNotFoundError: [Errno 2] No such file or directory: '/usr/local/share/jupyter/kernels/python3_delayed/run.shss': '/usr/local/share/jupyter/kernels/python3_delayed/run.shss'
[W 2021-06-21 15:51:28.582 ServerApp] Unhandled error
[E 2021-06-21 15:51:28.584 ServerApp] {
      "Host": "localhost:8888",
      "Connection": "keep-alive",
      "Content-Length": "126",
      "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.101 Safari/537.36",
      "X-Xsrftoken": "2|23c6813d|7801256b8d161d81ab7d6b004653b0e4|1622566377",
      "Authorization": "token 0bf1c62bc73200a5883091de47fc25593443229410f4e436",
      "Content-Type": "text/plain;charset=UTF-8",
      "Accept": "*/*",
      "Sec-Gpc": "1",
      "Origin": "http://localhost:8888",
      "Sec-Fetch-Site": "same-origin",
      "Sec-Fetch-Mode": "cors",
      "Sec-Fetch-Dest": "empty",
      "Referer": "http://localhost:8888/lab/workspaces/auto-o/tree/Untitled7.ipynb",
      "Accept-Encoding": "gzip, deflate, br",
      "Accept-Language": "en-US,en;q=0.9",
      "Cookie": "_xsrf=2|23c6813d|7801256b8d161d81ab7d6b004653b0e4|1622566377; username-localhost-8889=\"2|1:0|10:1624061638|23:username-localhost-8889|44:MmM3N2U5ODMwZTgwNDM3MDg4OTU2ZTJmZTFiMjhlNjU=|6e21168b1ddb7e6f41c8e249d9c42839b4b5c14cca39074e75d13f5e8fc4ebee\"; username-localhost-8888=\"2|1:0|10:1624315888|23:username-localhost-8888|44:ZDg5Y2UyOGMyNTExNGEzZGJkNGM2YmFkNzc5YjZiYzI=|6375141e2c35d6cd707a89f3a4ac957f607880ec2ed5462cbb8636665298f4f7\""
    }
[E 2021-06-21 15:51:28.584 ServerApp] 500 POST /api/sessions?1624315888539 (::1) 27.09ms referer=http://localhost:8888/lab/workspaces/auto-o/tree/Untitled7.ipynb
```
and the end-user sees this:
![image](https://user-images.githubusercontent.com/22599560/122838541-1a744480-d2ab-11eb-91a1-229120eef1c7.png)

As a result, I think you'd have some hoops to jump through from `jupyter_client`, up through the gateway server and over to the user's notebook/jupyter server.  That said, I don't want to discourage you from contributing a solution.

One thing that I suspect could make this much easier is if/when the [Kernel Provisioning](https://github.com/jupyter/jupyter_client/issues/608) proposal were to get done.  With this, a given provisioner (in this case, one that launches k8s-based kernels) could validate the "parameters" during the [`pre_launch()`](https://github.com/jupyter/jupyter_client/pull/612/files#diff-b370bb301ffe7b9de4edb6f77835f50f930c12552fdacd203839f9710daeedb5R144) hook and raise appropriate exceptions which should propagate back to the user a bit more cleanly.

**Comment by kevin-bates** [comment]
[_Kernel Provisioning_](https://jupyter-client.readthedocs.io/en/latest/provisioning.html) is available in jupyter_client 7.0 but, unfortunately, will not be available in EG until it support provisioners (as they collide with the current process proxies). This effort is slated for EG 4.0.

However, because this is a general framework issue, I'm closing this issue here.  We can revisit this down the road.
